### PR TITLE
Allow reading from ReaderWriter when there is data.

### LINKED ISSUE #336: ReaderWriter doesn't allow reads while the buffer isn't full
The read function doesn't work unless the buffer is full, or the writer closes the pipe.

### PR DESCRIPTION
**Pull request by floitsch**
Fixes #336.

### CODE REVIEW

File: lib/reader_writer.toit [hunk 1]
```diff
@@ -80,27 +77,38 @@ monitor ReaderWriterHelper_:
...
   read:
-    await: outgoing_ or writer_closed_
-    if outgoing_:
-      result := outgoing_
-      outgoing_ = null
+    await: buffer_ or writer_closed_
+    result := ?
+    if buffer_:
+      if fullness_ != buffer_.size:
+        result = buffer_[..fullness_]
```
Comment (erikcorry) [inline lib/reader_writer.toit:106]: I think this violates the expectation that read will return fresh byte arrays that can be held onto.  Subsequent calls to read will return arrays that alias each other.

Comment (floitsch) [reply lib/reader_writer.toit]: refactored.

File: lib/reader_writer.toit [hunk 2]
```diff
@@ -80,27 +76,25 @@ monitor ReaderWriterHelper_:
...
 
   read:
-    await: outgoing_ or writer_closed_
-    if outgoing_:
-      result := outgoing_
-      outgoing_ = null
+    await: fullness_ != 0 or writer_closed_
+    result := ?
+    if fullness_ != 0:
+      result = buffer_[..fullness_].copy
```
Comment (erikcorry) [inline lib/reader_writer.toit:96]: I think copy takes from and to, so you are not creating a slice object here for no reason.

Comment (floitsch) [reply lib/reader_writer.toit]: done.

**Pull request merged**